Sources/Fuzzilli/Core/ProgramBuilder.swift

```diff
@@ -343,10 +343,11 @@ public class ProgramBuilder {
         return variables.map(adopt)
     }
 
-    private func adoptTypes(from origInstr: Instruction, to newInstr: Instruction) {
+    private func adoptTypes(from origInstr: Instruction) {
+        let newInstr = program.lastInstruction
         for (originalVariable, adoptedVariable) in zip(origInstr.allOutputs, newInstr.allOutputs) {
-            if let type = typeMaps.last![originalVariable], type != .unknown {
-                program.setRuntimeType(of: adoptedVariable, to: type)
+            if let instrMap = typeMaps.last![originalVariable], let type = instrMap[origInstr.index], type != .unknown {
```

Comment:
IIUC, this currently does not make use of updated type information since it only looks at the instructions outputs when setting the type (and so, e.g. a type recollection on `obj.__proto__ = X` would not be processed here as `obj` is not an output)? But this probably also depends on the upcoming type information API refactoring, so I'm fine with leaving it this way for now

Reply:
Yes, that would currently fix it I think. But in the future, maybe we'd like to be able to update types of variables not directly affected by an instruction (e.g. if we are clever enough to detect changes to common prototypes or something like that). Then this still wouldn't be enough.
Maybe another way to implement it (possibly after the upcoming refactoring) is to iterate over the runtime type data structure by instruction index instead of by variable. This might require converting the data structure once at the start of an adoption though.

Reply:
hmm so the structure will have to be able to answer queries not only on variable but also on instruction, so it will be a bit more complex, so we just leave it to refactoring?

Reply:
yes, let's do that. It would also be an option to rebuild a different data structure for `adopt` that can efficiently be queried by instruction instead of by variable. Because I think that query is only needed during adoption. Not 100% sure though :)